File: backend/byteBlog/src/main/java/com/jeswin8801/byteBlog/service/webapp/abstracts/BlogService.java
package com.jeswin8801.byteBlog.service.webapp.abstracts;

import com.jeswin8801.byteBlog.entities.dto.GenericResponseDto;
import com.jeswin8801.byteBlog.entities.dto.MessageResponseDto;
import com.jeswin8801.byteBlog.entities.dto.blog.*;
import com.jeswin8801.byteBlog.entities.model.Blog;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;
import java.util.Map;
import java.util.Set;

public interface BlogService {
    GenericResponseDto<TagsDto> getAllTags();

    GenericResponseDto<MessageResponseDto> addNewBlog(
            PostBlogRequestDto postBlogRequestDto,
            MultipartFile markdownFile,
            List<MultipartFile> images
    );

    GenericResponseDto<Set<BlogsCompactResponseDto>> getAllBlogsByAuthor(String username);

    GenericResponseDto<BlogDto> getBlogByHeading(String headingUri);

    GenericResponseDto<AuthorDto> getBlogAuthorDetails(String username);

    Blog getBlogById(String blogId);

    GenericResponseDto<Map<String, Integer>> getAllUsedTags();

    GenericResponseDto<Set<BlogsCompactResponseDto>> getAllBlogsByTag(String tag);

    GenericResponseDto<BlogsCompactResponseDto> getFeaturedBlog();

    GenericResponseDto<Set<BlogsCompactResponseDto>> getLatestBlogsAsPageable(int pageNumber, int pageSize);
}

File: frontend/src/app/service/blog/blog.service.ts
import { HttpClient } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { Observable, tap } from 'rxjs';
import { environment } from '../../../environments/environment';
import { TagsDto } from '../../models/dtos/blog/tags-dto';
import { PostBlogDto } from '../../models/dtos/blog/post-blog-dto';
import { StandardResponseDto } from '../../models/dtos/user/standard-response-dto';
import { ObjectMapper } from 'json-object-mapper';
import { AppConstants } from '../../common/app.constants';

@Injectable({
  providedIn: 'root',
})
export class BlogService {
  constructor() {}

  private readonly http = inject(HttpClient);

  getAllTags(): Observable<TagsDto> {
    return this.http
      .get<TagsDto>(`${environment.apiUrl}/blog/get-all-tags`)
      .pipe(
        tap(() => {
          console.log('successfully retrieved blog tags');
        })
      );
  }

  addNewBlog(
    blog: PostBlogDto,
    markdownFile: File,
    images: FileList
  ): Observable<StandardResponseDto> {
    const formData: FormData = new FormData();
    formData.append(
      'blog',
      new Blob([ObjectMapper.serialize(blog) as string], {
        type: 'application/json',
      })
    );
    formData.append('markdown', markdownFile);

    if (images)
      for (let i = 0; i < images.length; i++)
        formData.append('images', images[i]);

    return this.http.post<StandardResponseDto>(
      `${environment.apiUrl}/blog`,
      formData
    );
  }

  getAllBlogsByUsername(username: string): Observable<any[]> {
    return this.http.get<any[]>(
      `${environment.apiUrl}/unrestricted/blog/author/get-all-blogs?username=` +
        username
    );
  }

  getBlogByHeading(headingUri: string): Observable<any> {
    return this.http.get<any[]>(
      `${environment.apiUrl}/unrestricted/blog/get-blog?heading-uri=` +
        headingUri
    );
  }

  getMarkdownFileFromUrl(url: string): Observable<string> {
    return this.http.get(url, {
      responseType: 'text',
    });
  }

  getAuthorDetails(username: string): Observable<any> {
    return this.http.get<any>(
      `${environment.apiUrl}/unrestricted/blog/author?username=` + username
    );
  }

  getAllUsedTags(): Observable<any> {
    return this.http.get<any>(
      `${environment.apiUrl}/unrestricted/blog/used-tags`
    );
  }

  getAllBlogsWithTag(tag: string): Observable<any> {
    return this.http.get<any>(
      `${environment.apiUrl}/unrestricted/blog/get-by-tag?tag=` + tag
    );
  }

  getFeaturedBlog(): Observable<any> {
    return this.http.get<any>(
      `${environment.apiUrl}/unrestricted/blog/featured`
    );
  }

  getLatestBlogsAsPageable(pageNumber: number): Observable<any> {
    return this.http.get<any>(
      `${environment.apiUrl}/unrestricted/blog/latest?page-number=` +
        pageNumber +
        `&page-size=` +
        AppConstants.PAGE_SIZE
    );
  }
}
